lib/teleterm/clusters/cluster_auth_test.go

```diff
@@ -0,0 +1,133 @@
+
+	// Test valid pin.
+	stream.mockSend = func(res *api.LoginPasswordlessResponse) error {
+		require.Equal(t, api.PasswordlessPrompt_PASSWORDLESS_PROMPT_PIN, res.Prompt)
+		return nil
+	}
+	stream.mockRecv = func() (*api.LoginPasswordlessRequest, error) {
+		return &api.LoginPasswordlessRequest{Request: &api.LoginPasswordlessRequest_Pin{Pin: "1234"}}, nil
+	}
+
+	prompt := newPwdlessLoginPrompt(context.Background(), stream)
+	pin, err := prompt.PromptPIN()
+	require.NoError(t, err)
+	require.Equal(t, "1234", pin)
+
+	// Test invalid pin.
+	stream.mockRecv = func() (*api.LoginPasswordlessRequest, error) {
+		return &api.LoginPasswordlessRequest{Request: &api.LoginPasswordlessRequest_Pin{Pin: ""}}, nil
+	}
+
+	_, err = prompt.PromptPIN()
+	require.True(t, trace.IsBadParameter(err))
+}
+
+func TestPwdlessLoginPrompt_PromptTouch(t *testing.T) {
+	stream := &mockLoginPwdlessStream{}
+
+	// Test count 0.
+	stream.mockSend = func(res *api.LoginPasswordlessResponse) error {
+		require.Equal(t, api.PasswordlessPrompt_PASSWORDLESS_PROMPT_TAP, res.Prompt)
+		return nil
+	}
+
+	prompt := newPwdlessLoginPrompt(context.Background(), stream)
+	err := prompt.PromptTouch()
+	require.NoError(t, err)
+
+	// Test count++.
+	stream.mockSend = func(res *api.LoginPasswordlessResponse) error {
+		require.Equal(t, api.PasswordlessPrompt_PASSWORDLESS_PROMPT_RETAP, res.Prompt)
+		return nil
+	}
+
+	err = prompt.PromptTouch()
+	require.NoError(t, err)
+	err = prompt.PromptTouch()
+	require.NoError(t, err)
+}
+
+func TestPwdlessLoginPrompt_PromptCredential(t *testing.T) {
+	stream := &mockLoginPwdlessStream{}
+
+	unsortedCreds := []*wancli.CredentialInfo{
+		{User: wancli.UserInfo{Name: "foo"}}, // will select
+		{User: wancli.UserInfo{Name: "bar"}},
+		{User: wancli.UserInfo{Name: "ape"}},
+		{User: wancli.UserInfo{Name: "llama"}},
+	}
+
+	// Test valid index.
+	stream.mockSend = func(res *api.LoginPasswordlessResponse) error {
+		require.Equal(t, api.PasswordlessPrompt_PASSWORDLESS_PROMPT_CREDENTIAL, res.Prompt)
+		require.Equal(t, []string{"ape", "bar", "foo", "llama"}, res.Usernames)
+		return nil
+	}
+	stream.mockRecv = func() (*api.LoginPasswordlessRequest, error) {
+		return &api.LoginPasswordlessRequest{Request: &api.LoginPasswordlessRequest_UsernameIndex{UsernameIndex: 2}}, nil
+	}
+
+	prompt := newPwdlessLoginPrompt(context.Background(), stream)
+	cred, err := prompt.PromptCredential(unsortedCreds)
+	require.NoError(t, err)
+	require.Equal(t, "foo", cred.User.Name)
+
+	// Test invalid index.
+	stream.mockRecv = func() (*api.LoginPasswordlessRequest, error) {
+		return &api.LoginPasswordlessRequest{Request: &api.LoginPasswordlessRequest_UsernameIndex{UsernameIndex: 4}}, nil
+	}
+	_, err = prompt.PromptCredential(unsortedCreds)
+	require.True(t, trace.IsBadParameter(err))
+}
+
+type mockLoginPwdlessStream struct {
+	grpc.ServerStream
+	mockSend func(resp *api.LoginPasswordlessResponse) error
+	mockRecv func() (*api.LoginPasswordlessRequest, error)
```

Comment:
Maybe:

```suggestion
		req *api.LoginPasswordlessRequest
		assertResp func(*api.LoginPasswordlessResponse)
```

so it's a bit simpler to use? (A bit subjective, though.)